### PR TITLE
2D U-Net model with Medical Decathlon Brain Tumor Dataset

### PR DESCRIPTION
**Pull request by tonyreina**
This is a 2D U-Net which has been trained on the Medical Decathlon dataset (http://medicaldecathlon.com/). The Medical Decathlon dataset has a commercially-friendly CC BY-SA 4.0 license (https://creativecommons.org/licenses/by-sa/4.0/) which means it can be repackaged and redistributed.  The PR includes a subset of the Brain Tumor Segmentation (BraTS) MRI data from Medical Decathlon.  The 2D U-Net model has an average Dice metric of > 0.85. The python script `mri_segmentation.py` will load the sample of validation MRI images and run inference on the OpenVINO model. It will save png files for the MRI, ground truth masks, and prediction masks.

### CODE REVIEW

**Comment by snosov1** [comment]
Hey, Tony! Thanks for the contribution! We'll review shortly and provide feedback.

@lachinov Dmitry, please, take a look

**Comment by tonyreina** [comment]
Thanks.  I also have a 3D version I can contribute for same dataset, but would need some help to make sure that the IE 3D operations were enabled.

**Comment by snosov1** [comment]
The next release (in March) will, actually, have the 3D sample - there's 2 models (one of which solves Brats). You can consider extending this sample once it gets here.

**Comment by yury-gorbachev** [comment]
That said, 3d operations are functional in OpenVINO since R5, it is just proper sample that was missing.

**Comment by snosov1** [comment]
Btw, @tonyreina, do you plan to release the training code for the model and the checkpoint in the original format? Reason is, this is the direction we're going for the set of models - replace the IR-only versions with their "trainable" counterparts.

For example, we're replacing [license-plate-recognition-barrier-0001](https://github.com/opencv/open_model_zoo/tree/master/intel_models/license-plate-recognition-barrier-0001) with the same model distributed as TF checkpoint https://github.com/opencv/openvino_training_extensions/tree/develop/tensorflow_toolkit/lpr

**Comment by tonyreina** [comment]
Yes. I am going through SDL currently to open source all of the code.

-Tony



Sent from my Verizon, Samsung Galaxy smartphone


-------- Original message --------

**Comment by tonyreina** [comment]
> That said, 3d operations are functional in OpenVINO since R5, it is just proper sample that was missing.

I'm able to use the model optimizer for R5 to create a 3D U-Net model for this, but the Python example (same one I used for this 2D U-Net but expanded to NCHWD) gives a segmentation fault.  I'm not sure what is wrong (https://github.com/NervanaSystems/topologies/blob/master/3D_UNet/keras_training_only_version/openvino_models/inference_openvino.py)

**Comment by tonyreina** [comment]
Just wanted to check and see if this model was accepted. Thanks.

**Comment by snosov1** [comment]
Hey, Tony!

Sorry for the delay. Thing is, your model is the first one to be contributed via GitHub and our infra is in the transition stage from internal repos/tools to GitHub. ATM, we can pull code contributions (i.e. demos/samples), but the models are a little bit trickier. One issue is that IRs have to be rebuilt with each release of model_optimizer (it produces backward-incompatible models, occasionally). 

To move this PR forward, I'd suggest

1. @lachinov, please, review the sample code
2. @tonyreina please, update us on the status of your SDL/open sourcing. I think, the best way would be to link to your repo and the model in the original format. If it takes too long, we can consider uploading your manually generated IR somewhere.

**Comment by tonyreina** [comment]
Thanks Sergei. Yes. We're starting the Protex scans for the repo to open source. I'm not sure of the timeline, but we've did the first review last week and have already spoken with legal. Hoping by end of April.

For the open source repo I was only planning to have the source code instead of the IR or saved model. It seemed like that was better to put in the model zoo. Is there any mechanism for updating an older IR to a newer format? The model only contains 2D convolution/transposed conv/ReLU/max pooling so I think it should be easy to maintain.

File: demos/python_demos/mri_segmentation.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    for idx in range(0, img_indicies.shape[0], batch_size):
+
+        res = exec_net.infer(inputs={input_blob:
+                                     input_data[idx:(idx+batch_size),
+                                     :n_channels]})
+
+        # Save the predictions to array
+        predictions[idx:(idx+batch_size),] = res[out_blob]
+
+    if idx != (len(img_indicies)-1):  # Partial batch left in data
```
Comment (lachinov) [inline demos/python_demos/mri_segmentation.py]: Is it correct behavior to miss a few slices? Please consider expanding the last batch to the full batch size with zeros

File: demos/python_demos/mri_segmentation.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+
+    """
+    Evaluate model with Dice metric
+    """
+    for idx in range(img_indicies.shape[0]):
+        dice = dice_score(predictions[idx,0,:,:], label_data[idx,0,:,:])
+        log.info("Image #{}: Dice score = {:.4f}".format(img_indicies[idx], dice))
+
+    if args.plot:
+        plot_predictions(predictions, input_data, label_data, img_indicies, args)
```
Comment (lachinov) [inline demos/python_demos/mri_segmentation.py]: Please produce output (preferably nifti) so it can be viewed and evaluated externally.

**Comment by snosov1** [comment]
> Hoping by end of April.

Ok, noted. Hopefully, we'll finalize the transition of our tools to open source by this time frame as well.

> Is there any mechanism for updating an older IR to a newer format?

Essentially, we plan to have the models stored in the original format *somewhere*. And then the plan is to call model optimizer after the download (it works this way for public models now). 

We have a place now - https://download.01.org . But we're not very satisfied with it (to say the least). So, we're looking for the alternative storage/hosting place, but the idea for the general workflow will be the same

> I think it should be easy to maintain.

Best/easiest maintenance is no/automatic maintenance =) The procedure I've described always uses the *current* model optimizer, so we avoid any possible incompatibilities. BTW, my understanding was that you'd need the model in original format to be able to re-train/fine-tune it with your code anyway, no?

File: demos/python_demos/mri_segmentation.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+import sys
+import os
+from argparse import ArgumentParser
+import numpy as np
+import logging as log
+from time import time
+from openvino.inference_engine import IENetwork, IEPlugin
+
+def dice_score(pred, truth):
```
[the remark this answers is not in the document]
Comment (tonyreina) [reply demos/python_demos/mri_segmentation.py]: The additive smooth is essential for protecting against the corner case of an empty mask and prediction (which should be Dice = 1). Refer to the paper https://arxiv.org/pdf/1707.03237.pdf or the section on additive/Laplace smoothing in Wikipedia (https://en.wikipedia.org/wiki/Additive_smoothing). For popular implementations, please refer to https://github.com/ellisdg/3DUnetCNN/blob/master/unet3d/metrics.py  or https://github.com/jakeret/tf_unet/blob/bd2c940f2f73c1b57ca2397d6d200c72f0e4e4bc/tf_unet/unet.py

[the remark this answers is not in the document]
Comment (lachinov) [reply demos/python_demos/mri_segmentation.py]: Additive smoothing is used in order to make this function differentiable. In return, this so-called Dice loss function (all of the above-stated references) is different from Dice index in cases with small gt masks (with the size of few pixels), and it adds bias to the final evaluation. As you stated, empty masks should produce Dice index equals to 1 indeed. However, we are not constrained by the differentiability requirement, thus we can handle empty masks separately with if-else statement.

File: demos/python_demos/mri_segmentation.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        predictions[idx:(idx+batch_size),] = res[out_blob]
+
+    if idx != (len(img_indicies)-1):  # Partial batch left in data
+        log.info("Partial batch left over in dataset.")
+
+    """
+    Evaluate model with Dice metric
+    """
+    for idx in range(img_indicies.shape[0]):
+        dice = dice_score(predictions[idx,0,:,:], label_data[idx,0,:,:])
```
[the remark this answers is not in the document]
Comment (tonyreina) [reply demos/python_demos/mri_segmentation.py]: A soft Dice score is perfectly valid as the metric.

[the remark this answers is not in the document]
Comment (lachinov) [reply demos/python_demos/mri_segmentation.py]: True, but it might be misleading (ex. all 0.49 predicts will produce non-zero evaluation and empty prediction mask) and produces higher values than Dice score. Since the data is taken from Medical Segmentation Decathlon, I suggest using the same validation scheme as they do.

File: demos/python_demos/mri_segmentation.py (hunk 1 above)
Comment (tonyreina) [reply demos/python_demos/mri_segmentation.py]: Good question.  Yes. I can do that instead.

File: demos/python_demos/mri_segmentation.py (hunk 2 above)
Comment (tonyreina) [reply demos/python_demos/mri_segmentation.py]: In line 52 I have the function "plot_predictions" which saves the images as pngs.  This is much easier than Nifti format since the casual user won't know how to the view a Nifti file and the advanced user will be able to save the NumPy to Nifti.

File: demos/python_demos/mri_segmentation.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+            row += 1
+
+def load_data():
+    """
+    Modify this to load your data and labels
+    """
+
+    # Load data
+    # You can create this Numpy datafile by running the create_validation_sample.py script
+    data_file = np.load("mri_segmentation_data/validation_data.npz")
```
[the remark this answers is not in the document]
Comment (tonyreina) [reply demos/python_demos/mri_segmentation.py]: That would require the additional steps of the user downloading the entire BraTS dataset (several GB) and then pre-processing the data just to pull out a few dozen 2D slices. Instead, it makes more sense to save those few dozen slices in a NumPy array (which is shipped in the model zoo like we do for car.bmp in the license plate identification model). I provide the Nifti processing scripts in the training repo (https://github.com/IntelAI/unet) for those that want to re-train the model or pull out other data, but as this is just a model zoo to show a few sample cases, it seemed like overkill to download several GB of data for 12 slices.

File: demos/python_demos/mri_segmentation.py [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+
+    return input_data, msks_data, img_indicies
+
+def load_model(fp16=False):
+    """
+    Load the OpenVINO model.
+    """
+    log.info("Loading U-Net model to the plugin")
+
+    if fp16:  # Floating point 16 is for Myriad X
```
[the remark this answers is not in the document]
Comment (tonyreina) [reply demos/python_demos/mri_segmentation.py]: Yes. I can remove. I also have a FP16 version of the model for use with MyriadX/NCS2.  I wasn't sure if you wanted it for the Model Zoo. I will remove the additional code. Let me know if you'd like to include the FP16 model for NCS2.

File: demos/python_demos/mri_segmentation.py [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+    infer_time = []
+
+    for i in range(args.number_iter):
+        t0 = time()
+        res = exec_net.infer(inputs={input_blob: input_data[0:batch_size,:n_channels]})
+        infer_time.append((time() - t0) * 1000)
+
+    average_inference = np.average(np.asarray(infer_time))
+    log.info("Average running time of one batch: {:.5f} ms".format(average_inference))
+    log.info("Images per second = {:.3f}".format(batch_size * 1000.0 / average_inference))
```
[the remark this answers is not in the document]
Comment (tonyreina) [reply demos/python_demos/mri_segmentation.py]: Sure. Makes sense. I'll change from images to slices.  As for linear performance, what is the proposed alternative? This is how it's been done in the TensorFlow CNN model benchmarks.

File: demos/python_demos/mri_segmentation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tonyreina) [reply demos/python_demos/mri_segmentation.py]: No. I don't agree.

File: demos/python_demos/mri_segmentation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tonyreina) [reply demos/python_demos/mri_segmentation.py]: That's a fair statement, but I think it is only relevant when we are trying to show that our model is somehow state of the art. Why is this an issue in the model zoo? Aren't we just showing examples of how to use OV? I'm just as much of an expert on this as you and we simply disagree here.

File: demos/python_demos/mri_segmentation.py [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+        else:
+            row += 1
+
+def load_data():
+    """
+    Modify this to load your data and labels
+    """
+
+    # Load data
+    # You can create this Numpy datafile by running the create_validation_sample.py script
```
[the remark this answers is not in the document]
Comment (tonyreina) [reply demos/python_demos/mri_segmentation.py]: Functionality is handled by including the NumPy test dataset along with the model. This way the user won't have to do anything else to run the script.

File: demos/python_demos/mri_segmentation.py (hunk 1 above)
Comment (tonyreina) [reply demos/python_demos/mri_segmentation.py]: Done. Thanks.

File: demos/python_demos/mri_segmentation.py [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+    if args.cpu_extension and "CPU" in args.device:
+        plugin.add_cpu_extension(args.cpu_extension)
+
+    # Read IR
+    # If using MYRIAD then we need to load FP16 model version
+    model_xml, model_bin = load_model(args.device == "MYRIAD")
+
+    log.info("Loading network files:\n\t{}\n\t{}".format(model_xml, model_bin))
+    net = IENetwork(model=model_xml, weights=model_bin)
+    #net = IENetwork.from_ir(model=model_xml, weights=model_bin) # Old API
```
[the remark this answers is not in the document]
Comment (tonyreina) [reply demos/python_demos/mri_segmentation.py]: Done. Thanks.

File: demos/python_demos/mri_segmentation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tonyreina) [reply demos/python_demos/mri_segmentation.py]: Removed. Thanks.

File: demos/python_demos/mri_segmentation.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tonyreina) [reply demos/python_demos/mri_segmentation.py]: Done. Thanks.

**Comment by tonyreina** [comment]
> Best/easiest maintenance is no/automatic maintenance =) The procedure I've described always uses the current model optimizer, so we avoid any possible incompatibilities. BTW, my understanding was that you'd need the model in original format to be able to re-train/fine-tune it with your code anyway, no?

Yes. That's correct. The GitHub repo for the training code is here:  https://github.com/IntelAI/unet
I have instructions in the repo to convert the model to OV:  https://github.com/IntelAI/unet/tree/master/2D/openvino

So that would be the mechanism if someone wanted to re-train/fine tune the model. I'm happy to also supply the Keras pre-trained model to the Model Zoo, but wasn't sure that was allowed. Just let me know. It's easy to add. I also have a FP16 model version for NCS2, but wasn't sure if that was desired either for the zoo.

File: demos/python_demos/mri_segmentation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lachinov) [reply demos/python_demos/mri_segmentation.py]: That wouldn't require additional steps since instead of the single numpy file we can ship single nifti series. This will make the whole pipeline (load data -> preprocess -> infer -> postprocess -> save) more clear.

File: demos/python_demos/mri_segmentation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (snosov1) [reply demos/python_demos/mri_segmentation.py]: Let me try to mediate =)

I guess, technically, the name of the function is misleading. If it were `dice_loss` then it would be, technically, correct. Now, I think, there's 2 options - either we rename the function or introduce `if'`s (as @lachinov suggests). In terms of implementation both are equally simple. 

Personally, I think, I'm leaning towards additional `if`s, but in the end, it's your call, @tonyreina . Either way should work.

File: demos/python_demos/mri_segmentation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (snosov1) [reply demos/python_demos/mri_segmentation.py]: I guess this is related to the naming. If you leave `dice_score`, then, it's, technically, incorrect and it needs to be changed. If you decide to go with `dice_loss`, then you can leave it as is. Fair enough?

File: demos/python_demos/mri_segmentation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (snosov1) [reply demos/python_demos/mri_segmentation.py]: @tonyreina I think, we can go with your logic about making the sample more friendly to "casual" users. So, plotting and output can, probably, stay as is. For input, though, I'd agree with @lachinov - it makes more sense to take "images" in the original format (i.e. make your conversion logic a part of this script). So, instead of storing a single ("sample") `.npz` you'd store a single series of `.nifti`s. This way it doesn't introduce any additional overhead (i.e. downloading of several GBs). Makes sense?

Btw, as for storing the Brats data itself - we, probably, need to take care about that. Both in terms of binaries in git and potential license issues. Thinking of options - can you include 1 example series (same 10Mb you have in the PR, but in nifti format) to the repo you're planning to release?

**Comment by vladimir-dudnik** [comment]
Jenkins please retry a build

**Comment by IRDonch** [comment]
@tonyreina If you're still up for it, we now have a more well-defined model contribution process: https://github.com/opencv/open_model_zoo/blob/develop/CONTRIBUTING.md. So please follow that.

I'm going to close this MR for now, but feel free to reopen it once you update your branch.

**Comment by snosov1** [comment]
@IRDonch FYI

There appeared to be licensing incompatibilities between OMZ and this model. We didn't pursue the addition further after we realized that. 

That said, things may have changed since then.
